core/src/main/java/nl/nn/adapterframework/pgp/PGPAction.java

```diff
@@ -0,0 +1,115 @@
+import nl.nn.adapterframework.configuration.ConfigurationException;
+import nl.nn.adapterframework.util.ClassUtils;
+import org.apache.commons.io.IOUtils;
+import org.bouncycastle.openpgp.PGPException;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URL;
+
+/**
+ * This is an abstraction of general pgp actions
+ * such as encryption, verification, etc.
+ * to be used for {@link nl.nn.adapterframework.pipes.PGPPipe}
+ *
+ * @author Murat Kaan Meral
+ */
+public abstract class PGPAction {
+
+	protected InMemoryKeyring keyringConfig;
+	private String[] publicKeys;
+	private String secretKey, secretPassword;
+
+	/**
+	 * A general constructor. It checks if any of the vars are null.
+	 *
+	 * @param publicKey      Array of path strings that point to public keys.
+	 * @param secretKey      String that contains the path to the secret key.
+	 * @param secretPassword String that contains the password for the secret key.
+	 * @param vars           Objects that should not be null for the action to be performed.
+	 * @throws ConfigurationException When any of the given vars is null.
+	 */
+	PGPAction(String[] publicKey, String secretKey, String secretPassword, Object... vars) throws ConfigurationException {
+		verifyNotNull(vars);
+		this.publicKeys = publicKey;
+		this.secretPassword = secretPassword;
+		this.secretKey = secretKey;
+	}
+
+	/**
+	 * Generates a keyring configuration with public keys and the private key.
+	 *
+	 * @throws ConfigurationException When the files do not exist, or unexpected PGP exception has occurred.
+	 */
+	public void configure() throws ConfigurationException {
+		try {
+			// Create configuration
+			KeyringConfigCallback callback = KeyringConfigCallbacks.withUnprotectedKeys();
+			if (secretPassword != null)
+				callback = KeyringConfigCallbacks.withPassword(secretPassword);
+
+			keyringConfig = KeyringConfigs.forGpgExportedKeys(callback);
+
+			// Add public keys
+			if (publicKeys != null) {
+				for (String s : publicKeys) {
+					URL url = ClassUtils.getResourceURL(Thread.currentThread().getContextClassLoader(), s);
+					keyringConfig.addPublicKey(IOUtils.toByteArray(url.openStream()));
+				}
+			}
+
+			// Add private key
+			if (secretKey != null) {
+				URL url = ClassUtils.getResourceURL(Thread.currentThread().getContextClassLoader(), secretKey);
+				keyringConfig.addSecretKey(IOUtils.toByteArray(url.openStream()));
+			}
+		} catch (IOException | PGPException e) {
+			throw new ConfigurationException("Unknown exception has occurred.", e);
+		}
+	}
+
+	/**
+	 * Runs the given action (which may be any extensions of this abstract class).
+	 *
+	 * @param inputStream Input for the action.
+	 * @return OutputStream that contains the encrypted/plaintext based on the action.
+	 * @throws Exception Any exception that can be thrown during the action.
+	 */
+	public abstract OutputStream run(InputStream inputStream) throws Exception;
```

Comment:
The result of this action should be an InputStream, not an OutputStream.
A caller can use an OutputStream result to write to, whereas it should read from this.
In all your implementations the OutputStream is in fact a ByteArrayOutputStream, that is just a wrapper around the result which is the byte array.